httpclient5/src/main/java/org/apache/hc/client5/http/socket/UnixDomainSocketFactory.java

```diff
@@ -0,0 +1,179 @@
+    private static final String JDK_UNIX_SOCKET_ADDRESS_CLASS = "java.net.UnixDomainSocketAddress";
+    private static final String JUNIXSOCKET_SOCKET_CLASS = "org.newsclub.net.unix.AFUNIXSocket";
+    private static final String JUNIXSOCKET_ADDRESS_CLASS = "org.newsclub.net.unix.AFUNIXSocketAddress";
+
+    private enum Implementation {
+        JDK,
+        JUNIXSOCKET,
+        NONE
+    }
+
+    private static final Implementation IMPLEMENTATION = detectImplementation();
+
+    private static Implementation detectImplementation() {
+        try {
+            Class.forName(JUNIXSOCKET_SOCKET_CLASS);
+            LOG.debug("Using JUnixSocket Unix Domain Socket implementation");
+            return Implementation.JUNIXSOCKET;
+        } catch (final ClassNotFoundException e) {
+            try {
+                Class.forName(JDK_UNIX_SOCKET_ADDRESS_CLASS);
+                LOG.debug("Using JDK Unix Domain Socket implementation");
+                return Implementation.JDK;
+            } catch (final ClassNotFoundException e2) {
+                LOG.debug("No Unix Domain Socket implementation found");
+                return Implementation.NONE;
+            }
+        }
+    }
+
+    /**
+     * Checks if Unix Domain Socket support is available.
+     *
+     * @return true if Unix Domain Socket support is available, false otherwise
+     */
+    public static boolean isAvailable() {
+        return IMPLEMENTATION != Implementation.NONE;
+    }
+
+    /**
+     * Default instance of {@link UnixDomainSocketFactory}.
+     */
+    private static final UnixDomainSocketFactory INSTANCE = new UnixDomainSocketFactory();
+
+    /**
+     * Gets the singleton instance of {@link UnixDomainSocketFactory}.
+     *
+     * @return the singleton instance
+     */
+    public static UnixDomainSocketFactory getSocketFactory() {
+        return INSTANCE;
+    }
+
+    public SocketAddress createSocketAddress(final Path socketPath) {
+        if (!isAvailable()) {
+            throw new UnsupportedOperationException("Unix Domain Socket support is not available");
+        }
+        Args.notNull(socketPath, "Unix domain socket path");
+
+        try {
+            if (IMPLEMENTATION == Implementation.JDK) {
+                // JDK implementation
+                final Class<?> addressClass = Class.forName(JDK_UNIX_SOCKET_ADDRESS_CLASS);
+                final Method ofMethod = addressClass.getMethod("of", Path.class);
+                return (SocketAddress) ofMethod.invoke(null, socketPath);
+            } else {
+                // JUnixSocket implementation
+                final Class<?> addressClass = Class.forName(JUNIXSOCKET_ADDRESS_CLASS);
+                final Method ofMethod = addressClass.getMethod("of", Path.class);
+                return (SocketAddress) ofMethod.invoke(null, socketPath);
+            }
+        } catch (final ReflectiveOperationException ex) {
+            throw new RuntimeException("Could not create UDS SocketAddress", ex);
```

Comment:
@rschmitt  Is `RuntimeException` here too generic? Would it be clearer to wrap this in an `IOException` (or an `UncheckedIOException`) instead?

Reply:
If we get past the `isAvailable()` check, then this exception should never throw.